src/base/algodiff/owl_algodiff_ops_builder.ml

```diff
@@ -347,4 +347,85 @@ module Make (Core : Owl_algodiff_core_sig.Sig) = struct
         b
     in
     f
+
+
+  module type Aiso = sig
+    val label : string
+    val ff : t array -> t
+    val df : int array -> t -> t array -> t array -> t
+    val dr : int array -> t array -> t -> t ref -> (t * t) list
+  end
+
+  let build_aiso =
+    let build_info =
+      Array.fold_left
+        (fun (i, t, m, idxs) x ->
+          match m, x with
+          | _, F _ | _, Arr _ -> succ i, t, m, idxs
+          | `n, DR (_, _, _, _, t', _) -> succ i, t', `r, [ i ]
+          | `f, DR (_, _, _, _, t', _) ->
+            if t' > t
+            then succ i, t', `r, []
+            else if t' = t
+            then failwith "clash"
```

Comment:
Here and below should we be more informative?

Reply:
I've changed it to `error: forward and backward clash at the same level` as above